### PR TITLE
feat: Refactor global state

### PR DESCRIPTION
**Pull request by hirunatan**

### CODE REVIEW

**Comment by watchoutfreedom** [comment]
Amazing XD 

Some notes:

-Calling NextJS Router directly on pages seems to be enough for us now, but it's true that getting dynamics routes for networks seems complex and may be better to have it separatedd. Do you think about any other case apart from network related routing that we may need this?

-I think this way of getting everything together is great, do you think it's good if we replicate the folder structure of other folders, or follow the naming for models (like Buttons instead of Explore) what do you think about the structure or concept behind  the naming?

**Comment by albjeremias** [comment]
i'm making a branch to make the buttons form to work.. because this branch will break it!

**Comment by hirunatan** [comment]
> -Calling NextJS Router directly on pages seems to be enough for us now, but it's true that getting dynamics routes for networks seems complex and may be better to have it separatedd. Do you think about any other case apart from network related routing that we may need this?

I've reviewing this, and yes, we don't need anything strange. It should be enough to use directly the Router from components and events. The only advantage of having the Router.push inside an event is that it's executed asynchronously, not in the middle of the execution of other event. But the side effects will usually be handled by suscriptions to the Router events.

For dynamic routes the Router.push is enough: [https://nextjs.org/docs/api-reference/next/router#usage](https://nextjs.org/docs/api-reference/next/router#usage)


> -I think this way of getting everything together is great, do you think it's good if we replicate the folder structure of other folders, or follow the naming for models (like Buttons instead of Explore) what do you think about the structure or concept behind the naming?

This is a good question, and I have not a definitive answer. I think the structure of the state module should not reflect the data types (Buttons, Networks, etc.) This is for the services, that fits the back API. State should reflect more the organization of the application screens. But not one by one, perhaps some pages could share one submodule of state (for example, Login, Signup and Profile pages can use User submodule). And some submodules (e.g. Networks) may be used from many pages.

File: src/state/Routes.tsx [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import Router from 'next/router';
+
+import { EffectEvent } from 'store/Event';
+import { GlobalState } from 'store/Store';
+
+// TODO: not clear if this should be an event, a service
+//       or directly call to Router in pages. Think about it.
+
+export class NavigateTo implements EffectEvent {
```
Comment (hirunatan) [inline src/state/Routes.tsx:9]: I think we could remove this event and replace it with a direct call to Router.push.

**Comment by albjeremias** [comment]
Doubts:
 1- on fieldTags (https://github.com/helpbuttons/hb-front/blob/develop/src/elements/Fields/FieldTags/index.tsx) I've tried to use
https://react-hook-form.com/api/usefieldarray but got very confusing because useForm mostly uses <input and or FieldTags is storing the tags not in a input field. Also 

2- FileUpload (https://github.com/helpbuttons/hb-front/blob/develop/src/elements/Fields/FieldImagesUpload/index.tsx) have no idea how-to handle this with useForm. there exists this Controller (https://react-hook-form.com/get-started#IntegratingwithUIlibraries) but i can't get it to work

3- Location (https://github.com/helpbuttons/hb-front/blob/develop/src/elements/Fields/FieldLocation/index.tsx) in here we need to register 3 different fields, latitude, longitude and radius... should we register multiple ? Or store it as an array or object? the api expects 3 different fields.

File: src/state/Routes.tsx (hunk 1 above)
Comment (watchoutfreedom) [reply src/state/Routes.tsx:9]: If we don't use the async feature yes. But you know better now how do we need the async for routes

**Comment by albjeremias** [comment]
Ok, I got the FieldTags and FieldImagesUpload to work. 

On the fieldTags I got this error, which gets me really confused.
```
Warning: Cannot update a component (`ButtonNew`) while rendering a different component (`FieldTags`). To locate the bad setState() call inside `FieldTags`, follow the stack trace as described in https://reactjs.org/link/setstate-in-render
FieldTags@webpack-internal:///./src/elements/Fields/FieldTags/index.tsx:25:15
div
form
Form@webpack-internal:///./src/elements/Form/index.tsx:12:18
```

**Comment by watchoutfreedom** [comment]
> > -Calling NextJS Router directly on pages seems to be enough for us now, but it's true that getting dynamics routes for networks seems complex and may be better to have it separatedd. Do you think about any other case apart from network related routing that we may need this?
> 
> 
> 
> I've reviewing this, and yes, we don't need anything strange. It should be enough to use directly the Router from components and events. The only advantage of having the Router.push inside an event is that it's executed asynchronously, not in the middle of the execution of other event. But the side effects will usually be handled by suscriptions to the Router events.
> 
> 
> 
> For dynamic routes the Router.push is enough: [https://nextjs.org/docs/api-reference/next/router#usage](https://nextjs.org/docs/api-reference/next/router#usage)
> 
> 
> 
> 
> 
> > -I think this way of getting everything together is great, do you think it's good if we replicate the folder structure of other folders, or follow the naming for models (like Buttons instead of Explore) what do you think about the structure or concept behind the naming?
> 
> 
> 
> This is a good question, and I have not a definitive answer. I think the structure of the state module should not reflect the data types (Buttons, Networks, etc.) This is for the services, that fit the back API. State should reflect more the organization of the application screens. But not one by one, perhaps some pages could share one submodule of state (for example, Login, Signup and Profile pages can use User submodule). And some submodules (e.g. Networks) may be used from many pages.

Copy . I get both points. In which scenario do you think Dynamic async routing can be helpful?

**Pull request merged**